File: 6.DP/BOJ/(1932번)정수 삼각형/[jeunghun] solution.swift
let n = Int(readLine()!)!
var data = Array<Array<Int>>()
var dp = (1..<n).map { i in
    Array(repeating: 0, count: i)
}

for _ in 0..<n {
    data.append(readLine()!.split(separator: " ").map { Int($0)! })
}

dp.append(data.last!)

for i in 0..<n - 1 {
    for j in 0..<n - i - 1 {
        dp[n - 2 - i][j] = max(dp[n - 1 - i][j] + data[n - 2 - i][j], dp[n - 1 - i][j + 1] + data[n - 2 - i][j])
    }
}

print(dp[0][0])

File: 6.DP/BOJ/(1932번)정수 삼각형/[jeunghun] solution.py
n = int(input())
data = []
dp = [[0] * i for i in range(1, n)]

for _ in range(n):
    data.append(list(map(int, input().split())))

dp.append(data[-1])

for i in range(n - 2, -1, -1):
    for j in range(i + 1):
        dp[i][j] = max(dp[i + 1][j] + data[i][j], dp[i + 1][j + 1] + data[i][j])

print(dp[0][0])